src/main/java/org/jenkinsci/plugins/docker/workflow/WithContainerStep.java

```diff
@@ -171,16 +171,27 @@ private static FilePath tempDir(FilePath ws) {
         private static final long serialVersionUID = 1;
         private final String container;
         private final String[] envHost;
+        private final String ws;
 
-        Decorator(String container, EnvVars envHost) {
+        Decorator(String container, EnvVars envHost, String ws) {
             this.container = container;
             this.envHost = Util.mapToEnv(envHost);
+            this.ws = ws;
         }
 
-        @Override public Launcher decorate(Launcher launcher, Node node) {
+        @Override public Launcher decorate(final Launcher launcher, Node node) {
             return new Launcher.DecoratedLauncher(launcher) {
                 @Override public Proc launch(Launcher.ProcStarter starter) throws IOException {
                     List<String> prefix = new ArrayList<String>(Arrays.asList("docker", "exec", container, "env"));
+                    if (ws != null) {
+                        FilePath cwd = starter.pwd();
+                        if (cwd != null) {
+                            String path = cwd.getRemote();
+                            if (!path.equals(ws)) {
+                                launcher.getListener().getLogger().println("JENKINS-33510: working directory will be " + ws + " not " + path);
```

Comment:
I considered simply calling something like

```java
prefix.add("cd " + path + " && ");
```

but it will not work since the array here is expected to be a parsed argv, which from a `sh` step will be `nohup`, then `sh`, then `-c`, then a long metascript. Now we could wrap all of that in _another_ `sh -c …`, but then you need to quote all the original commands, which is not a simple task since those commands already use some tricky quotes.

Or we could use [this trick](https://github.com/ndeloof/docker-slaves-plugin/blob/c8a3559ac808ee7f3cf490a7b1d1c6d382251bd7/trampoline/src/trampoline/main.go), which I am hesitant to do.